packages/cardano-services-client/src/WebSocket.ts

```diff
@@ -0,0 +1,182 @@
+    this.url = cfg.url;
+
+    this.networkInfoSubjects = {
+      eraSummaries: new ReplaySubject<EraSummary[]>(1),
+      genesisParameters: new ReplaySubject<Cardano.CompactGenesis>(1),
+      ledgerTip: new ReplaySubject<Cardano.Tip>(1),
+      lovelaceSupply: new ReplaySubject<SupplySummary>(1),
+      protocolParameters: new ReplaySubject<Cardano.ProtocolParameters>(1),
+      stake: new ReplaySubject<StakeSummary>(1)
+    };
+    this.networkInfo = this.networkInfoSubjects;
+
+    this.networkInfoProvider = {
+      eraSummaries: () => firstValueFrom(this.networkInfo.eraSummaries),
+      genesisParameters: () => firstValueFrom(this.networkInfo.genesisParameters),
+      healthCheck: () => Promise.resolve({ ok: this.status === 'connected' }),
+      ledgerTip: () => firstValueFrom(this.networkInfo.ledgerTip),
+      lovelaceSupply: () => firstValueFrom(this.networkInfo.lovelaceSupply),
+      protocolParameters: () => firstValueFrom(this.networkInfo.protocolParameters),
+      stake: () => firstValueFrom(this.networkInfo.stake)
+    };
+
+    this.connect();
+  }
+
+  private connect() {
+    this.status = 'connecting';
+
+    const ws = new WebSocket(this.url);
+
+    // eslint-disable-next-line sonarjs/cognitive-complexity
+    ws.onmessage = (event) => {
+      try {
+        if (typeof event.data !== 'string') throw new Error('Unexpected data from WebSocket ');
+
+        const message = NJSON.parse<WSMessage>(event.data);
+        const { clientId, networkInfo } = message;
+
+        if (clientId) this.logger.info(`Connected with clientId ${(this.clientId = clientId)}`);
+
+        if (networkInfo) {
+          const { eraSummaries, genesisParameters, ledgerTip, lovelaceSupply, protocolParameters, stake } = networkInfo;
+
+          if (eraSummaries) this.networkInfoSubjects.eraSummaries.next(eraSummaries);
+          if (genesisParameters) this.networkInfoSubjects.genesisParameters.next(genesisParameters);
+          if (lovelaceSupply) this.networkInfoSubjects.lovelaceSupply.next(lovelaceSupply);
+          if (protocolParameters) this.networkInfoSubjects.protocolParameters.next(protocolParameters);
+          if (stake) this.networkInfoSubjects.stake.next(stake);
+          // Emit ledgerTip as last one
+          if (ledgerTip) this.networkInfoSubjects.ledgerTip.next(ledgerTip);
+        }
+      } catch (error) {
+        this.logger.error(error, 'While parsing message', event.data, this.clientId);
+      }
+    };
+
+    ws.onclose = () => {
+      this.logger.info('WebSocket client connection closed', this.clientId);
+      this.clientId = NOT_CONNECTED_ID;
+      this.retry();
+    };
+
+    ws.onerror = (error: unknown) => {
+      this.logger.error(isEventError(error) ? error.error : error, 'Async error from WebSocket client', this.clientId);
+      ws.close();
+    };
+
+    ws.onopen = () => {
+      this.status = 'connected';
+      this.ws = ws;
+      this.heartbeat();
+    };
+  }
+
+  private heartbeat() {
```

Comment:
Dumb question: what is the purpose of heartbeat? I'd expect tip notifications from the server to keep the connection alive :thinking:

Reply:
It's purpose is on the reverse side.
Exactly thanks to the tip there is no need for a server to client heartbeat.
The purpose of this is to let the server know the client is still connected (and not disconnected with some hard disconnections which leaves the connection "open" at TCP-IP level but broken anyway).

Reply:
Hm, are you sure it's required? I haven't worked with ws for a very long time, but I'd expect such failures to be handled by the protocol or lower level libraries

Reply:
It's more a nice to have, we had some public discussion on this point as well.
It's true that underlaying layers **should have** the seme kind of logic, anyway each o pin the route has its own timeout...

Reply:
LW-10841